pkg/controller/deployment/sync.go

```diff
@@ -541,22 +541,25 @@ func (dc *DeploymentController) cleanupDeployment(oldRSs []*extensions.ReplicaSe
 	if deployment.Spec.RevisionHistoryLimit == nil {
 		return nil
 	}
-	diff := int32(len(oldRSs)) - *deployment.Spec.RevisionHistoryLimit
+
+	// Avoid deleting replica set with deletion timestamp set or with non-zero replica counts
+	cleanablefilter := func(rs *extensions.ReplicaSet) bool {
+		return rs != nil && rs.ObjectMeta.DeletionTimestamp == nil && rs.Status.Replicas == 0 && *(rs.Spec.Replicas) == 0 && rs.Generation <= rs.Status.ObservedGeneration
+	}
+	cleanableRSes := controller.FilterReplicaSets(oldRSs, cleanablefilter)
+
+	diff := int32(len(cleanableRSes)) - *deployment.Spec.RevisionHistoryLimit
```

Comment:
I think this slightly changes the meaning of revisionHistoryLimit. If I have a Deployment (revisionHistoryLimit=1) with two old ReplicaSets and a new one:
d.replicas=3 with rs-1.replicas=0 (old), rs-2.replicas=1 (old), and rs-3.replicas=2 (new)

With the old behavior we would cleanup rs-1, with the behavior you have here we won't until the rollout finishes. I think revisionHistoryLimit should take all old controllers into account and then filter out those it cannot delete as we already do. Maybe prefiltering just for deletionTimestamp is enough.